Use NSViewController.present() with custom animator #108

hazeycode opened the pull request:
While this still changes the contentViewController, presenting view controllers behave exactly like a stack. Works flawlessly (in my testing) with dismiss:

Fixes #108


<!-- Issuehunt content -->

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#108: Use `NSViewController.present()` with custom animator](https://issuehunt.io/repos/119822304/issues/108)
---
</details>
<!-- /Issuehunt content-->

sindresorhus commented:
Thanks for working on this.

You have a lot of trailing whitespace. Make sure `swiftlint` passes without any warnings.

sindresorhus commented:
Cancelling a conversion no longer correctly works:

<img width="472" alt="Screen Shot 2021-05-06 at 16 14 02" src="https://user-images.githubusercontent.com/170270/117273176-1371ae00-ae86-11eb-8c96-7389b6f63a08.png">

hazeycode commented:
Sorry about that @sindresorhus, I have fixed the cancellation state bug that I introduced and tested the conversion state cancelations and completions more thoroughly.

sindresorhus commented on Gifski/ConversionViewController.swift:
```diff
@@ -116,19 +118,15 @@ final class ConversionViewController: NSViewController {
 
 	private func didComplete(conversion: Gifski.Conversion, gifUrl: URL) {
 		let conversionCompleted = ConversionCompletedViewController(conversion: conversion, gifUrl: gifUrl)
-		stopConversion { [weak self] in
```
Why are you removing `weak`?

hazeycode replied:
This was an incorrect fix for a bug where the delayed tooltip on ConversionCompletedViewController could throw an error under unlikely user behaviour. Steps to reproduce:

1. Start a conversion
2. Hit escape exactly during the transition from ConversionViewController to ConversionCompletedViewController (spamming escape as soon as transition begins) 
3. Click back button

I have found a correct fix for this and will update shortly

sindresorhus commented on Gifski/EditVideoViewController.swift:
```diff
@@ -108,9 +106,7 @@ final class EditVideoViewController: NSViewController {
-		let videoDropController = VideoDropViewController()
-		push(viewController: videoDropController)
-		AppDelegate.shared.previousEditViewController = nil
+		self.presentingViewController?.dismiss(self)
```
You don't need `.self` here.

sindresorhus commented on Gifski/Utilities.swift:
```diff
@@ -3001,45 +3001,89 @@ extension URL {
+extension NSViewController {
+	func push(viewController: NSViewController, completion: (() -> Void)? = nil) {
+		let animator = ViewControllerPresentationAnimator(completion: completion)
+		self.present(viewController, animator: animator)
```
You don't need `self`.

hazeycode commented on Gifski/ConversionCompletedViewController.swift:
```diff
@@ -70,6 +70,10 @@ final class ConversionCompletedViewController: NSViewController {
 				return
 			}
 
+			guard self.view.window?.contentViewController == self else {
+				return
+			}
+
 			self.tooltip.show(from: self.draggableFile, preferredEdge: .maxY)
```
This is the fix for the tooltip crash. It's an odd one and quite difficult to reproduce.

sindresorhus commented on Gifski/ConversionViewController.swift:
```diff
@@ -116,12 +118,8 @@ final class ConversionViewController: NSViewController {
-			viewController.isConverting = false
-
-			self?.push(viewController: viewController)
+		stopConversion { [self] in
```
Same as https://github.com/sindresorhus/Gifski/pull/241#discussion_r627336411

hazeycode replied:
This one is a little different. The strong reference is required here so that the closure keeps ConversionViewController from being released during the transition.
There might be a better way that I haven't found yet to ensure this without requiring user code to keep a ref

hazeycode replied:
Must be the change in ownership model, before all view controllers ownership traced back to the edit view controller held by App

sindresorhus commented on Gifski/ConversionCompletedViewController.swift:
```diff
@@ -192,10 +196,7 @@ final class ConversionCompletedViewController: NSViewController {
-		let viewController = AppDelegate.shared.previousEditViewController!
-		viewController.isConverting = false
-		push(viewController: viewController)
+		self.presentingViewController?.presentingViewController?.dismiss(self)
```
You don't need `self` here.

sindresorhus commented:
If you drag and drop a video, then drag and drop another video onto the editor, and then press "Cancel", you end up with this:

<img width="917" alt="Screen Shot 2021-05-07 at 00 18 52" src="https://user-images.githubusercontent.com/170270/117339400-e3e49500-aec9-11eb-87b9-c35d98760b7b.png">

sindresorhus commented:
This may or may not be the same issue as the above: If you drag and drop a video, convert it, and on the success screen, drag and drop another video, and then press Cancel, you end up with:

<img width="472" alt="Screen Shot 2021-05-07 at 00 22 44" src="https://user-images.githubusercontent.com/170270/117339794-58b7cf00-aeca-11eb-85bf-e6e288fceca6.png">

sindresorhus commented:
Have you tested this change on macOS 10.15 or 10.14?

sindresorhus commented on Gifski/ConversionCompletedViewController.swift:
```diff
@@ -70,6 +70,10 @@ final class ConversionCompletedViewController: NSViewController {
 				return
 			}
 
+			guard self.view.window?.contentViewController == self else {
```
I feel like this is just hiding a different issue. `self` is already weak at this point, so the only way this can happen is if this controller is retained while not being the current one.

hazeycode replied:
This is true. There seems to be some time during the transition that this can be invoked but before ARC has cleanup up the popped VC

sindresorhus replied:
If there's really no better solution than this, it should at least have a code comment about this.

hazeycode replied:
I'm not sure why viewDidAppear is getting called at all on this view controller after clicking back. Will do some more digging

hazeycode commented on Gifski/MainWindowController.swift:
```diff
@@ -121,8 +124,10 @@ final class MainWindowController: NSWindowController {
 
-		let editController = EditVideoViewController(inputUrl: inputUrl, asset: asset, videoMetadata: videoMetadata)
-		window?.contentViewController?.push(viewController: editController)
+		rootVideoDropViewController.popAll {
+			let editController = EditVideoViewController(inputUrl: inputUrl, asset: asset, videoMetadata: videoMetadata)
+			self.rootVideoDropViewController.push(viewController: editController)
+		}
```
Fix for https://github.com/sindresorhus/Gifski/pull/241#issuecomment-833701277

hazeycode commented on Gifski/VideoDropViewController.swift:
```diff
@@ -56,7 +56,13 @@ final class VideoDropViewController: NSViewController {
+			return
+		}
+
+		popAll {
+			let editController = EditVideoViewController(inputUrl: inputUrl, asset: asset, videoMetadata: videoMetadata)
+			rootVideoDropViewController.push(viewController: editController)
+		}
```
Fix for https://github.com/sindresorhus/Gifski/pull/241#issuecomment-833703219

hazeycode commented:
I only have macOS version 11.2.3 for testing at this time

sindresorhus commented on Gifski/Utilities.swift:
```diff
@@ -3001,44 +3001,78 @@ extension URL {
+		present(viewController, animator: animator)
+	}
+
+	func popAll(completion: (() -> Void)? = nil) {
```
Can you follow the `UINavigationController` naming? `popToRootViewController(animated:)`

sindresorhus commented:
If I drag in a video, then drag another one into the editor, the editor animates slightly in size. This is new behavior. It should only fade between the view. Try out the App Store version to see the behavior difference.

---

You have some swiftlint warnings.

sindresorhus commented on Gifski/ConversionCompletedViewController.swift:
```diff
@@ -192,10 +198,7 @@ final class ConversionCompletedViewController: NSViewController {
-		let viewController = AppDelegate.shared.previousEditViewController!
-		viewController.isConverting = false
-		push(viewController: viewController)
+		presentingViewController?.presentingViewController?.dismiss(self)
```
Use `.pop()` here too for consistency?

sindresorhus commented on Gifski/MainWindowController.swift:
```diff
@@ -121,8 +124,10 @@ final class MainWindowController: NSWindowController {
-		window?.contentViewController?.push(viewController: editController)
+		rootVideoDropViewController.popToRootViewController(animated: false) {
+			let editController = EditVideoViewController(inputUrl: inputUrl, asset: asset, videoMetadata: videoMetadata)
+			self.rootVideoDropViewController.push(viewController: editController)
```
Should `self` be weak here?

sindresorhus commented on Gifski/Utilities.swift:
```diff
@@ -3001,44 +3001,82 @@ extension URL {
+		guard let window = view.window else {
+			return
+		}
+		guard let rootVC = (window.windowController as? MainWindowController)?.rootVideoDropViewController else {
```
You are mixing reusable utilities with app-specific logic.

sindresorhus commented:
Can you use the Xcode memory graph to ensure there are no retain cycles?

https://medium.com/zendesk-engineering/ios-identifying-memory-leaks-using-the-xcode-memory-graph-debugger-e84f097b9d15

I'm a little bit paranoid that this change introduces subtle retain cycles. I spent years fixing retain cycles in Gifski.

hazeycode commented:
I think moving the transition code out of the NSViewController extension in Utilities into a RootViewController would resolve the issues here. RootViewController would take responsibilty for the lifetimes and transitions between view controllers and would be the only window contentViewController.

sindresorhus commented:
> I think moving the transition code out of the NSViewController extension in Utilities into a RootViewController would resolve the issues here.

Why would it have to be moved out the extension. We could still have a root view controller that manages everything, while having the reusable transition logic in the extension.

hazeycode commented:
> Why would it have to be moved out the extension? We could still have a root view controller that manages everything, while having the reusable transition logic in the extension.

I guess it doesn't need to be, so long as we have a way to tell which view controller is root, which should be straight forward if the window's contentViewController is always the root view controller. I'll have another look at this when I've got some more time

hazeycode commented:
> Why would it have to be moved out the extension? We could still have a root view controller that manages everything, while having the reusable transition logic in the extension.

I guess it doesn't have to be. The problem is relying on NSViewController and it's friends to keep things alive, so far my attempts have resulted in reference cycles or things getting released too early.

sindresorhus commented:
I'm closing this. This is taking forever and the current code works fine. This was just meant to clean up the code, but it's really not worth the risk of breaking things more. We'll eventually move the whole app to SwiftUI at some point anyway.
